### ISSUE TITLE
Lowlevel pruning with reader

### ISSUE THREAD

**Issue by xor-gate** [ludocode/mpack#30]
Currently I extract a key : <map> to feed to a other piece of the API decoder and want to calculate the size in bytes of the map and the begin address. Here a example:

```
{ 
   "somekey" : 1,
   "anotherkey" : "value",
   "interestedkey" : {
       "subkey" : "awesomevalue",
       "subkey1" : true
    },
    "foo" : "bar"
}
```

I itterate over the parent map and search for "interestedkey", then I pass the reader context to a function which calculates `const char *buffer, size_t size` for a new reader. 

Currently I do in the extraction function:
```
static void di_rpc_deserialize_msg_params(mpack_reader_t *reader, struct di_rpc_msg *msg)
{
        const size_t keys = mpack_expect_map(reader);

        if (mpack_reader_error(reader) != mpack_ok) {
                msg->params.begin = NULL;
                msg->params.end   = NULL;
                return;
        }

        msg->params.begin = &reader->buffer[reader->pos - 1];

        for (size_t n = 0; n < keys; n++) {
                mpack_discard(reader);
                mpack_discard(reader);
        }

        msg->params.end = &reader->buffer[reader->pos];

        mpack_done_map(reader);
}
```

Only it feels very inefficient to itterate over all key-values and discard them. Is it valid to just rewind the reader context `pos` one byte (so it points to the map byte) and then discard from this point?

Like this:
```

static void di_rpc_deserialize_msg_params(mpack_reader_t *reader, struct di_rpc_msg *msg)
{
        mpack_expect_map(reader);

        if (mpack_reader_error(reader) != mpack_ok) {
                msg->params.buffer = NULL;
                msg->params.size   = 0;
                return;
        }

        reader->pos -= 1;
        msg->params.buffer = &reader->buffer[reader->pos];
        mpack_discard(reader);
        msg->params.size = (size_t)((const char *)&reader->buffer[reader->pos] - msg->params.buffer);
        mpack_done_map(reader);
}
```

Maybe lowlevel pruning would be nice to have in the reader API? This is especially useful in systems where you want to grab a chunk of mpack data and pass to another layer.

**Comment by ludocode** [comment]
Hmm. Well, rewinding by one byte is probably not the right idea, since the map is not guaranteed to be one byte. It might be three or even five bytes. The reader doesn't really have any concept of rewinding since values can be encoded in different ways, and the standard doesn't require any sort of canonical representation. So the number 5 could be represented as `05`, or `CC 05`, or `D1 00 05`, and so on. Rewinding it would require knowing how it was originally encoded.

If you're reading from a fixed buffer, you could rewind in a hackish way by recording `pos` and `left` before calling `mpack_expect_map()`, then reset them to their original values. You have to set both because they both change when bytes are read. (I've tried changing the implementation to store `pos` and `count` so that only one of them would change, but it was slower for some reason.) You would have to disable tracking for this though.

But what you're doing in the first bit of code, just looping and calling discard twice, is fine. This is actually the same thing as what the discard function itself does:

https://github.com/ludocode/mpack/blob/master/src/mpack/mpack-reader.c#L680-L685

So rewinding to discard in one shot won't really help, since it would be doing exactly the same thing.

The bigger problem with discard is that `mpack_skip_bytes()` is horribly implemented right now. It just contains a TODO to make it efficient. I haven't gotten around to fixing it yet (I didn't think anyone would be using it.) It actually needs three different implementations: one where there is no `fill` callback (so it can just set the internal data pointers), one with a `skip` callback to seek files, and then finally the current implementation if there is only a `fill` function for an unseekable stream. I'll be doing this today.

**Comment by xor-gate** [comment]
A more clean approach would be to "mpack_read_peek" or "mpack_read_tag_ahead" if the it is a map then discard with the mpack API function. But currently when I use `mpack_read_tag` the reader points already on the map and need to discard all the kv myself and `mpack_done_map`.

**Comment by ludocode** [comment]
Ah yes, I've wanted to implement an `mpack_peek_tag()` function for a while, and it would also be useful in some of my other projects. The problem here is that the mpack reader supports buffer sizes smaller than the size of all possible tags, so it wouldn't be possible to peek because fill would have to be called and there would be nowhere to store the data. There are in fact stress tests for reading and writing tags across such small buffer sizes to make sure fill and flush are working correctly across any amount of straddling data: 

https://github.com/ludocode/mpack/blob/af55d10691f5e2f71242c24905a9e8070e92a476/test/test-buffer.c#L95-L107

Of course I don't actually believe there's value in preserving this. Requiring a minimum buffer size of, say, 32 bytes would be totally fine. But there are other problems in introducing this. I did this with the writer for code size and performance reasons and there were some very bad side effects, which I reverted before releasing 0.7.1. It also makes fill a lot more complicated, since if there isn't enough data it has to move the existing data around and only partly fill the buffer.

`mpack_read_bytes_inplace()` already does this though so maybe the implementations could be shared. And `mpack_peek_tag()` could also be implemented with an assert on a minimum buffer size, so you would only require a minimum buffer size if you call this function. This is the solution I was originally leaning towards, but now I'm more likely go with just a global minimum buffer size across all functions. In any case, you've convinced me of the need for `mpack_peek_tag()`. I'll spend more time to think about it and aim to implement it in 0.8.
